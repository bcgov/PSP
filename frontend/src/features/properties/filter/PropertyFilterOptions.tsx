import { InputGroup, Select } from 'components/common/form';
import { useFormikContext } from 'formik';
import React from 'react';

import { IPropertyFilter } from './IPropertyFilter';

interface IPropertyFilterOptions {
  disabled?: boolean;
  options?: { label: string; value: string }[];
  placeholders?: Record<string, string>;
}

/**
 * Provides a dropdown with list of search options for properties.
 */
export const PropertyFilterOptions: React.FC<IPropertyFilterOptions &
  React.HTMLAttributes<HTMLElement>> = ({ disabled, options, placeholders, ...rest }) => {
  const state: {
    options: { label: string; value: string }[];
    placeholders: Record<string, string>;
  } = {
    options: options ?? [
      { label: 'PID/PIN', value: 'pid' },
      { label: 'Address', value: 'address' },
    ],
    placeholders: placeholders ?? {
      address: 'Enter an address',
      pid: 'Enter a PID or PIN',
    },
  };

  // access the form context values, no need to pass props
  const {
    values: { searchBy },
    setFieldValue,
  } = useFormikContext<IPropertyFilter>();
  const desc = state.placeholders[searchBy] || '';

  const reset = () => {
    setFieldValue(options ? options[0].value : 'pid', '');
  };

  return (
    <InputGroup
      fast={false}
      formikProps={null as any}
      prepend={
        <Select field="searchBy" options={state.options} onChange={reset} disabled={disabled} />
      }
      field={searchBy}
      placeholder={desc}
      disabled={disabled}
      {...rest}
    ></InputGroup>
  );
};
