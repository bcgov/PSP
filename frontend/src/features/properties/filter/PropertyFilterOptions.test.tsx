import { render } from '@testing-library/react';
import { useFormikContext } from 'formik';

import { PropertyFilterOptions } from './PropertyFilterOptions';

jest.mock('formik');

(useFormikContext as jest.Mock).mockReturnValue({
  values: {
    includeAllProperties: false,
  },
  setFieldValue: jest.fn(),
});
it('renders correctly...', () => {
  const { asFragment } = render(<PropertyFilterOptions />);
  expect(asFragment()).toMatchSnapshot();
});

it('disabled prop works as intended...', () => {
  const { container } = render(<PropertyFilterOptions disabled />);
  const searchBy = container.querySelector('select[name="searchBy"]');
  expect(searchBy).toHaveAttribute('disabled');
});
