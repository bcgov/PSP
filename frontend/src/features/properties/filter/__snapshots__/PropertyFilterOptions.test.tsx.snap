// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly... 1`] = `
<DocumentFragment>
  <div
    class="input-group"
  >
    <div
      class="input-group-prepend"
    >
      <div
        class="form-group"
      >
        <select
          class="form-select form-control"
          id="input-searchBy"
          name="searchBy"
        >
          <option
            class="option"
            value="pid"
          >
            PID/PIN
          </option>
          <option
            class="option"
            value="address"
          >
            Address
          </option>
        </select>
      </div>
    </div>
    <div
      class="input-group-content"
    >
      <div
        class="input form-group"
      >
        <input
          class="form-control"
          id="input-undefined"
          placeholder=""
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
